### PR TITLE
[RDY] Use libvirt bindings instead of parsing output

### PR DESCRIPTION
**Pull request by teto**
A bit of a WIP. 
When I try deploying my example, I get 
```
libvirt: Network Driver error : Network not found: no network with matching name 'default'
error: Network not found: no network with matching name 'default'
```
but my example.nix is very thin.

### CODE REVIEW

**Comment by teto** [comment]
the 'default' network in libvirtd is visible only by root. How nixops can use the network in such conditions ? Is it supposed to ? Genuine question, due to bugs, I never reached the functional stage with libvirtd and nixops :(

**Comment by teto** [comment]
I've been trying to install the changes with an overlay aka `nix-env -iA nixops-local ` but it fails when building the wheel:
```
/tmp/nix-build-nixops-1.6.2.drv-0/nixops/dist /tmp/nix-build-nixops-1.6.2.drv-0/nixops
nixops-_version_-py2-none-any.whl is not a valid wheel filename
```
It seems like the       `version='@version@',` is not substituted in my case. Is there anything special to do ?

Here is my overlay
```
   nixops-local = super.nixopsUnstable.overrideAttrs( oldAttrs: rec {

     # version = "2017-09-24";
     version = "1.6.2";
     name = "nixops-${version}";
     src = super.lib.cleanSource ~/nixops;
   });
```

**Comment by teto** [comment]
So release.nix generates a tarball that is used in the derivation. It seems not straightforward to use the overlay so I stayed with ./dev-shell and succeeded making the branch define and run the vm. I will do some more testing as I need some specific libvirt tools (like possibly running my own kernel/initramfs)

**Comment by teto** [comment]
This is getting better (aka work better).
I would like to convert some parts for python 3. Is there any work in progress for python ? or is it ok as long as it remains python2 compatible ?

**Comment by bjornfor** [comment]
About python 3 support: https://github.com/NixOS/nixops/issues/504

**Comment by teto** [comment]
I've removed unrelated changes and cleaned the code so it would be cool to have some other feedback. I've been using this without problems.

File: nixops/backends/libvirtd.py [hunk 1]
```diff
@@ -181,22 +202,19 @@ def iface(n):
...
-            ip_with_subnet = lines.split()[i + 2]
-            return ip_with_subnet.split('/')[0]
+        """
+        return ip
+        """
+        # alternative is VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_LEASE
+        ifaces = self.dom.interfaceAddresses(libvirt.VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_LEASE, 0)
+        if (ifaces == None):
+            self.log("Failed to get domain interfaces")
+            sys.exit(0)
```
Comment (bjornfor) [inline nixops/backends/libvirtd.py]: Shouldn't this be exit(1) or something? Also, the current code doesn't call exit() at all. I think adding a comment (in the commit message) about why adding a process exit point in this function is a good thing. (I.e. is it a bug in the current code that it does _not_ call exit?)

Comment (teto) [reply nixops/backends/libvirtd.py]: that's an oversight from my side, I wanted to put it back to "None" as in the original code and forgot it.

File: nixops/backends/libvirtd.py [hunk 2]
```diff
@@ -222,18 +243,19 @@ def start(self):
             self.private_ipv4 = self._parse_ip()
         else:
             self.log("starting...")
-            self._logged_exec(["virsh", "-c", "qemu:///system", "start", self.vm_id])
+            self.dom.create()
```
Comment (bjornfor) [inline nixops/backends/libvirtd.py]: Oh, not "start"? Weird API.

**Comment by bjornfor** [comment]
I don't have rights to merge, but this looks good to me. (But see comments.)

**Comment by bjornfor** [comment]
Please squash commits.

**Comment by teto** [comment]
I think the squashing can be done on the github side ? if not I can do it ofc

File: nixops/backends/libvirtd.py (hunk 2 above)
Comment (teto) [reply nixops/backends/libvirtd.py]: yes it's weird https://www.redhat.com/archives/libvir-list/2010-June/msg00140.html

**Comment by domenkozar** [comment]
@teto is this ready for review?

**Comment by teto** [comment]
@domenkozar it's ready, it depends on https://github.com/NixOS/nixpkgs/pull/30950#issuecomment-342510936 though.

**Comment by teto** [comment]
since I,ve rebased on master I can't destroy domains anymore. That's weird, I might try against a virtualbox to see if it's nixops or this Pr problem

**Comment by teto** [comment]
seems unrelated in fact, if a deploiyment is not deployed yet, apparently I should use delete instead of destroy. I haven't changed the provider behavior so this must be a generic nixops heavior, unrelated to the libvirt provider.

**Comment by teto** [comment]
I added a commit to check some returned values.

**Comment by teto** [comment]
Anything I can do to help with merge ? I've some other libvirt improvements in queue :)

**Comment by bjornfor** [comment]
It seems the fixup commit hasn't been squashed yet.

**Comment by teto** [comment]
I mistakenly deleted the remote branch so I am forced to open a new PR https://github.com/NixOS/nixops/pull/783 which is this one rebased & squashed.

**Pull request closed without merging**